Segment-Optimizely-X/Classes/SEGOptimizelyXIntegration.m

```diff
@@ -0,0 +1,192 @@
+            [self.client track:payload.event userId:self.userId eventTags:payload.properties];
+            SEGLog(@"[optimizely track:%@ userId:%@ eventTags:%@]", payload.event, self.userId, payload.properties);
+        }
+    }
+
+    NSString *segmentAnonymousId = [self.analytics getAnonymousId];
+    if (!trackKnownUsers && self.userTraits.count > 0) {
+        [self.client track:payload.event userId:segmentAnonymousId attributes:self.userTraits eventTags:payload.properties];
+        SEGLog(@"[optimizely track:%@ userId:%@ attributes:%@ eventTags:%@]", payload.event, segmentAnonymousId, self.userTraits, payload.properties);
+    } else {
+        [self.client track:payload.event userId:segmentAnonymousId eventTags:payload.properties];
+        SEGLog(@"[optimizely track:%@ userId:%@ eventTags:%@]", payload.event, segmentAnonymousId, payload.properties);
+    }
+}
+
+- (void)reset
+{
+    if ([self.manager getOptimizely] == nil) {
+        return;
+    } else {
+        [[NSNotificationCenter defaultCenter] removeObserver:self.observer];
+        SEGLog(@"[NSNotificationCenter defaultCenter] removeObserver:%@", self.observer);
+    }
+}
+
+#pragma mark - Experiment Viewed For NSNotification
+
+- (void)experimentDidGetViewed:(NSNotification *)notification
+{
+    OPTLYExperiment *experiment = notification.userInfo[OptimizelyNotificationsUserDictionaryExperimentKey];
+    OPTLYVariation *variation = notification.userInfo[OptimizelyNotificationsUserDictionaryVariationKey];
+    NSDictionary *properties = @{
+        @"experimentId" : [experiment experimentId],
+        @"experimentName" : [experiment experimentKey],
+        @"variationId" : [variation variationId],
+        @"variationName" : [variation variationKey]
+    };
+
+    // Trigger event as per our spec https://segment.com/docs/spec/ab-testing/
+
+    [self.analytics track:@"Experiment Viewed" properties:properties options:@{
+        @"integrations" : @{
+            @"Optimizely X" : @NO
+        }
+    }];
+}
+
+#pragma mark - Private - Queueing
+
+- (void)enqueueAction:(SEGTrackPayload *)payload
+{
+    @synchronized(payload)
+    {
+        SEGLog(@"%@ Optimizely not initialized. Enqueueing action: %@", self, payload);
+        @try {
+            //TO DO: Should we set a condition to stop queuing events?
+            // Should set timer? Android has 5 minutes before stopping
+            if (self.queue.count > 100) {
+                // Remove the oldest element.
+                [self.queue removeObjectAtIndex:0];
+                SEGLog(@"%@ removeObjectAtIndex: 0", self.queue);
+            }
+            [self setupTimer];
+            [self.queue addObject:payload];
+            SEGLog(@"Queue length %i", self.queue.count);
+        }
+        @catch (NSException *exception) {
+            SEGLog(@"%@ Error writing payload: %@", self, exception);
+        }
+    }
+}
+
+- (NSMutableArray *)queue
+{
+    if (!_queue) {
+        _queue = [@[] mutableCopy];
```

Comment:
Use `NSMutableDictionary dictionaryWithCapacity`

Reply:
Hmm, what would define the capacity in this case?

Reply:
Either do 0 (what you have right now), something reasonably small so it doesn't take up too much space (like 10), or upto 100 (whatever your max limit we decide for the queue).